stablehlo/dialect/TypeInference.cpp

```diff
@@ -3565,13 +3651,75 @@ LogicalResult verifyDotGeneralOp(std::optional<Location> location, Value lhs,
     return failure();
 
   auto inferredShape = inferredReturnShapes[0];
-  auto resultType = result.getType().cast<ShapedType>();
-  if (inferredShape.hasRank() && resultType.hasRank() &&
+  auto resultShapeType = result.getType().cast<ShapedType>();
+  if (inferredShape.hasRank() && resultShapeType.hasRank() &&
       failed(verifyCompatibleShape(inferredShape.getDims(),
-                                   resultType.getShape())))
-    return emitOptionalError(
-        location, "inferred shape '", dimSizesToString(inferredShape.getDims()),
-        "' ", "is incompatible with return type of operation ", resultType, "");
+                                   resultShapeType.getShape())))
+    return emitOptionalError(location, "inferred shape '",
+                             dimSizesToString(inferredShape.getDims()), "' ",
+                             "is incompatible with return type of operation ",
+                             resultShapeType, "");
+
+  Type lhsType = lhs.getType();
+  Type rhsType = rhs.getType();
+  Type resultType = result.getType();
+  llvm::SmallVector<Type, 3> typeEntries{lhsType, rhsType, resultType};
+  if (noneQuantized<quant::QuantizedType>(typeEntries)) return success();
+  // dot_general_c14
+  if (!allQuantized<quant::QuantizedType>(typeEntries)) {
+    return emitOptionalError(location,
+                             "not all of operands and result are quantized");
+  }
+  auto lhsQType = getElementTypeOrSelf(lhsType).cast<quant::QuantizedType>();
+  auto rhsQType = getElementTypeOrSelf(rhsType).cast<quant::QuantizedType>();
+  auto resultQType =
+      getElementTypeOrSelf(resultType).cast<quant::QuantizedType>();
+  // dot_general_c15
+  if (lhsQType.getStorageType() != rhsQType.getStorageType())
+    return emitOptionalError(location, "mismatched operand storage types, lhs ",
+                             lhsQType.getStorageType(), " and rhs ",
+                             rhsQType.getStorageType());
+  // dot_general_c16
+  auto expressedType = lhsQType.getExpressedType();
+  if (expressedType != rhsQType.getExpressedType() ||
+      expressedType != resultQType.getExpressedType())
+    return emitOptionalError(location,
+                             "mismatched operands and result expressed types");
+
+  auto rhsQPAType = rhsQType.dyn_cast<quant::UniformQuantizedPerAxisType>();
+  // dot_general_c17
+  auto expectedZP = 0;
```

Comment:
Can w do some thing this instead

```
  auto rhsQPAType = rhsQType.dyn_cast<quant::UniformQuantizedPerAxisType>();
SmallVector<int64_t> zeroPoints = rhsQPAType ? rhsQPAType.getZeroPoints() : {rhsQType.cast<quant::UniformQuantizedType>().getZeroPoint()};

if(llvm::all_of(zeroPoints, [](int64_t zp) {return zp != 0})) {
return emitOptionalError(location, "Zero point of rhs type ",rhsQType, " is expected to be 0");
}
```

Reply:
neat! needed additional minor change.  thank you.

we decided to drop `dot_general` verifier from this PR due to internal test failures. Removed this change for now. I will park it as a draft PR.